soci/gzip_index.go

```diff
@@ -0,0 +1,141 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package soci
+
+// #cgo CFLAGS: -I${SRCDIR}/../c/
+// #cgo LDFLAGS: -L${SRCDIR}/../out -lindexer -lz
+// #include "indexer.h"
+// #include <stdlib.h>
+// #include <stdint.h>
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type GzipIndex struct {
+	index *C.struct_gzip_index
+}
+
+// NewGzipIndex creates a new instance of `GzipIndex` from index byte blob on zTOC.
+func NewGzipIndex(indexData []byte) (*GzipIndex, error) {
+	index := C.blob_to_index(unsafe.Pointer(&indexData[0]))
+	if index == nil {
+		return nil, fmt.Errorf("cannot convert blob to gzip_index")
+	}
+	return &GzipIndex{
+		index: index,
+	}, nil
+}
+
+// NewGzipIndexFromFile creates a new instance of `GzipIndex` given gzip file name and span size.
+func NewGzipIndexFromFile(gzipFile string, spanSize int64) (*GzipIndex, error) {
+	cstr := C.CString(gzipFile)
+	defer C.free(unsafe.Pointer(cstr))
+
+	var index *C.struct_gzip_index
+	ret := C.generate_index(cstr, C.off_t(spanSize), &index)
+	if int(ret) < 0 {
+		return nil, fmt.Errorf("could not generate gzip index. gzip error: %v", ret)
+	}
+
+	return &GzipIndex{
+		index: index,
+	}, nil
+}
+
+// Close calls `C.free` on the pointer to `C.struct_gzip_index`.
+func (i *GzipIndex) Close() {
+	if i.index != nil {
+		C.free(unsafe.Pointer(i.index))
+	}
+}
+
+// IndexToBlob returns the byte slice containing the index.
+func (i *GzipIndex) IndexToBlob() ([]byte, error) {
```

Comment:
I would just rename this method to `Bytes`.

Reply:
Done.